documentation/InductorTritonTemplateKernel.md

```diff
@@ -0,0 +1,221 @@
+# How to let Torchinductor codegen from user-defined templates
+
+This tutorial assumes the reader has an overview of how Torchinductor codegen Triton code and a basic understanding of Triton languange. This tutorial is for developers who want to further improve the performance of some operations, like Convolution and MatMul, by manually-written triton kernel.
+
+## Motivation
+For most of the operations, Torchinductor is able to automatically codegen corresponding Triton kernel. However,
+for `ir.ExternKernel` in torchinductor/, Torchinductor usually call extern functions like `aten.convolution` for Convolution, `aten.mm.out` for MatrixMultiply, etc. But if you have better implementation of those operations in Triton, say `triton.ops.matmul`, one way to let Torchindcutor to use it is replace the kernel and codegen function call to `triton.ops.matmul`. But still, this is an ExternKernel call where the Torchinductor optimize the whole graph unware of what's happening inside the function call. Specifically, it could not benefit from epilog fusion that could alleviate the memory-bound bottlenecks. To get additional speedup beside the customized code, user could add an Triton template for Torchinductor as a reference to cedegen the target kernel.
+
+## Overview
+Codegen from custom templates is derived from codegen normal Triton kernels. So in this tutorial, we will only focus on the difference.
+The main python file for codegen triton templates is [triton_template.py](https://github.com/pytorch/torchdynamo/blob/main/torchinductor/codegen/triton_template.py). Other files that need to modify when adding a new implementation as template file for Torchinductor codegen are [scheduler.py](https://github.com/pytorch/torchdynamo/blob/main/torchinductor/scheduler.py), [ir.py](https://github.com/pytorch/torchdynamo/blob/main/torchinductor/ir.py).
+
+Class TemplateSchedulerNode is a subclass of BaseSchedulerNode. During the `init` of Scheduler, it will create TemplateSchedulerNode if the node `should_use_template()`, and append to the `self.nodes` list. And when Sheduler `codegen()`, the TemplateSchedulerNode will call `codege_template_call()`, where call [`template_codegen()`](https://github.com/pytorch/torchdynamo/blob/main/torchinductor/codegen/triton_template.py#L299) inside the function.
```

Comment:
```suggestion
Class TemplateSchedulerNode is a subclass of BaseSchedulerNode. During the `init` of Scheduler, it will create TemplateSchedulerNode if the node `should_use_template()`, and append to the `self.nodes` list. And when the Scheduler does `codegen()`, the TemplateSchedulerNode will call [`template_codegen()`](https://github.com/pytorch/torchdynamo/blob/main/torchinductor/codegen/triton_template.py#L299) to generate the combined kernel.
```